drivers/storage/azure/storage/azure_storage.go

```diff
@@ -0,0 +1,914 @@
+
+	id, ok := context.InstanceID(ctx)
+	if !ok || id == nil {
+		return nil, goof.New("Can't detach volume outside of Azure instance")
+	}
+	vmName := id.ID
+
+	fields := map[string]interface{}{
+		"provider": d.Name(),
+		"vmName":   vmName,
+		"volumeID": volumeID,
+	}
+
+	volume, err := d.getVolume(ctx, volumeID, types.VolumeAttachmentsRequested)
+	if err != nil {
+		return nil, goof.WithFieldsE(fields, "failed to get volume", err)
+	}
+	if len(volume.Attachments) == 0 {
+		return nil, errVolAlreadyDetached
+	}
+
+	vm, err := d.getVM(ctx, vmName)
+	if err != nil {
+		return nil, goof.WithFieldsE(fields, "failed to detach volume to VM", err)
+	}
+
+	disks := *vm.StorageProfile.DataDisks
+	for i, disk := range disks {
+		if disk.Name != nil && *disk.Name == volumeID {
+			// found the disk
+			disks = append(disks[:i], disks[i+1:]...)
+			break
+		}
+	}
+	newVM := armCompute.VirtualMachine{
+		Location: vm.Location,
+		VirtualMachineProperties: &armCompute.VirtualMachineProperties{
+			StorageProfile: &armCompute.StorageProfile{
+				DataDisks: &disks,
+			},
+		},
+	}
+
+	_, err = mustSession(ctx).vmClient.CreateOrUpdate(d.resourceGroup, vmName, newVM, nil)
+	if err != nil {
+		return nil, goof.WithFieldsE(fields, "failed to detach volume", err)
+	}
+
+	volume, err = d.getVolume(ctx, volumeID, types.VolumeAttachmentsRequested)
+	if err != nil {
+		return nil, goof.WithFieldsE(fields, "failed to get volume", err)
+	}
+	if len(volume.Attachments) != 0 {
+		return nil, goof.WithFieldsE(fields, "volume is not detached", err)
+	}
+	return volume, nil
+}
+
+// Snapshots returns all volumes or a filtered list of snapshots.
+func (d *driver) Snapshots(
+	ctx types.Context,
+	opts types.Store) ([]*types.Snapshot, error) {
+	// TODO Snapshots are not implemented yet
+	return nil, types.ErrNotImplemented
+}
+
+// SnapshotInspect inspects a single snapshot.
+func (d *driver) SnapshotInspect(
+	ctx types.Context,
+	snapshotID string,
+	opts types.Store) (*types.Snapshot, error) {
+	// TODO Snapshots are not implemented yet
+	return nil, types.ErrNotImplemented
+}
+
+// SnapshotCopy copies an existing snapshot.
+func (d *driver) SnapshotCopy(
+	ctx types.Context,
+	snapshotID, snapshotName, destinationID string,
+	opts types.Store) (*types.Snapshot, error) {
+	// TODO Snapshots are not implemented yet
+	return nil, types.ErrNotImplemented
+}
+
+// SnapshotRemove removes a snapshot.
+func (d *driver) SnapshotRemove(
+	ctx types.Context,
+	snapshotID string,
+	opts types.Store) error {
+	// TODO Snapshots are not implemented yet
+	return types.ErrNotImplemented
+}
+
+// Get volume or snapshot name without config tag
+func (d *driver) getPrintableName(name string) string {
+	return strings.TrimPrefix(name, d.tag()+azure.TagDelimiter)
+}
+
+// Prefix volume or snapshot name with config tag
+func (d *driver) getFullName(name string) string {
+	if d.tag() != "" {
+		return d.tag() + azure.TagDelimiter + name
+	}
+	return name
+}
+
+// Retrieve config arguments
```

Comment:
Hi @Andrey-mp,

I'm not going to make this comment on each of the below functions, but this remark does apply to all of the code below related to retrieving configuration properties.

There is no need to check the environment variables directly. Please see this [example](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/ebs/storage/ebs_storage.go#L1321-L1332) for the EBS driver (ignore the backwards-compatability checks). The [gofig](https://github.com/akutz/gofig) package uses @spf13's [Viper](https://github.com/spf13/viper), which in turn enables the assignment of environment variables to defined configuration properties. 

Let's look at [line \#87](https://github.com/cloudscaling/libstorage/blob/f4bf3809fe1e354dbccd7ef05086f06677c7ad6b/drivers/storage/azure/azure.go#L87) in your `azure.go`:

```go
r.Key(gofig.String, "", "", "", ConfigAZURESubscriptionIDKey)
```

In addition to notifying Gofig about the property `azure.subscriptionID`, the above function call instructs Viper to bind the property `azure.subscriptionID` to the CLI flag `--azureSubscriptionID` and the environment variable `AZURE_SUBSCRIPTIONID`. You can actually use Gofig to overwrite which CLI flag or environment variable the property is bound via the [`Key` function's trailing variadic parameter](https://github.com/akutz/gofig/blob/master/gofig_reg.go#L56-L114). 

For example, if we wanted to still use the config property `azure.subscriptionID` but change the CLI flag to `azureSubID` and the environment variable to `AZURE_SUB_ID` we could rewrite the above line like so:

```go
r.Key(gofig.String, "", "", "", ConfigAZURESubscriptionIDKey, "azureSubID", "AZURE_SUB_ID")
```

Once the configuration properties are bound to a CLI flag or environment variable, invoking `config.GetString(azure.ConfigAZURESubscriptionIDKey)` checks first the environment variable, CLI flag argument, and finally a configuration property in that order.